File: examples/python/best-practices/at-least-once-semantics.py
@durable_step
def calculate_total(step_context: StepContext, items: list) -> float:
    return sum(item["price"] for item in items)

@durable_execution
def lambda_handler(event: dict, context: DurableContext) -> float:
    # Safe to run multiple times - same input produces same output
    total = context.step(calculate_total(event["items"]))
    return total

File: examples/java/best-practices/at-least-once-semantics.java
// Safe to run multiple times - same input produces same output
var total = ctx.step("calculate_total", Double.class, stepCtx -> {
    List<Map<String, Object>> items = (List<Map<String, Object>>) input.get("items");
    return items.stream()
        .mapToDouble(item -> ((Number) item.get("price")).doubleValue())
        .sum();
});
return total;
